runelite-client/src/main/java/net/runelite/client/plugins/cannon/CannonPlugin.java

```diff
@@ -300,15 +327,20 @@ public void onChatMessage(ChatMessage event)
 		{
 			cannonPlaced = true;
 			addCounter();
+			justPlaced = true;
```

Comment:
From testing in game, it seems like it's not possible to add the furnace but also not load it if you have cballs in your inventory. So I think you could just check if there are >= 30 cballs here and if so set the cballs left to 30 then. Then you could avoid the whole inv snapshot thing. That would fix most of the scenarios except if you had a partially-filled broken cannon and also <30 cannon balls to re-load it when replacing.